### PR TITLE
feat: pix2pix example

### PR DESCRIPTION
**Pull request by drh-determined-ai**
## Description

Implement the pix2pix TensorFlow Keras GAN example.

## Test Plan

Run automated tests in `examples/tests/test_official.py`, `e2e_tests/tests/nightly/test_capability.py`, and `e2e_tests/tests/nightly/test_distributed.py`.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: examples/gan/pix2pix_tf_keras/data.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        return _preprocess_images(i, height, width, jitter, mirror)
+
+    ds = ds.map(_prep, num_parallel_calls=tf.data.experimental.AUTOTUNE)
+    return ds
+
+
+def main():
+    import matplotlib.pyplot as plt
+
+    PATH = download("http://efrosgans.eecs.berkeley.edu/pix2pix/datasets/", "facades")
```
Comment (azhou-determined) [inline examples/gan/pix2pix_tf_keras/data.py:128]: should we grab this URL from the configuration `yaml` instead?

File: examples/gan/pix2pix_tf_keras/export.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import matplotlib.pyplot as plt
+
+from determined import keras
+from determined.experimental import client
+
+from data import download, load_dataset
+
+
+def generate_and_plot_images(generator: tf.keras.Sequential) -> None:
+    path = download("http://efrosgans.eecs.berkeley.edu/pix2pix/datasets/", "facades")
```
Comment (azhou-determined) [inline examples/gan/pix2pix_tf_keras/export.py:16]: same here

Comment (drh-determined-ai) [reply examples/gan/pix2pix_tf_keras/export.py:16]: Since this code is meant to be run after a Determined experiment, your point is more important that it is with `data`'s `main`. If we're going to make your suggested change to either, we might as well do it to both, so I'll get on it! :-)

File: examples/gan/pix2pix_tf_keras/data.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import pathlib
```
[the remark this answers is not in the document]
Comment (drh-determined-ai) [reply examples/gan/pix2pix_tf_keras/data.py:1]: Thank you!

File: examples/gan/pix2pix_tf_keras/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+* **`pix2pix/sampling.py`**: Defines the "upsampling" and "downsampling" patterns used in the discriminator and generator.
+* **`data.py`**: Code for loading and preprocessing the data.
+* **`export.py`**: Exports a training checkpoint and uses it to generate images.
+* **`fit.py`**: Trains the Keras model _without_ using Determined.
+* **`model_def.py`**: Wraps the model, optimizers, and datasets using Determined and provides the entrypoint.
+* **`print_models.py`**: Generates images showing the layers structure of the discriminator and generator.
+
+### Configuration Files
+* **`const.yaml`**: Train the model with constant hyperparameter values.
+* **`distributed.yaml`**: Same as `const.yaml`, but use multiple GPUs for distributed training.
```
[the remark this answers is not in the document]
Comment (drh-determined-ai) [reply examples/gan/pix2pix_tf_keras/README.md:22]: Oops. I totally forgot about that. On it!

File: examples/gan/pix2pix_tf_keras/model_def.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+from determined.keras import InputData, TFKerasTrial, TFKerasTrialContext
+
+from pix2pix import Pix2Pix, make_discriminator_optimizer, make_generator_optimizer
+from data import download, load_dataset
+
+
+class Pix2PixTrial(TFKerasTrial):
+    def __init__(self, context: TFKerasTrialContext) -> None:
+        self.context = context
+        self.path = download(
```
[the remark this answers is not in the document]
Comment (drh-determined-ai) [reply examples/gan/pix2pix_tf_keras/model_def.py:12]: This is something I've been wondering about. I'll probably ask you for a bit more guidance in Slack.

File: examples/gan/pix2pix_tf_keras/data.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    real_image,
+    height,
+    width,
+    jitter=0,
+    mirror=False,
+) -> Tuple[tf.Tensor, tf.Tensor]:
+    if jitter > 0:
+        # Resizing to 286x286
+        input_image, real_image = _resize(
+            input_image, real_image, height + jitter, width + jitter
```
[the remark this answers is not in the document]
Comment (drh-determined-ai) [reply examples/gan/pix2pix_tf_keras/data.py:75]: I'm not sure I understand what you're getting at. I just copied this bit of preprocessing from the TF tutorial and haven't thought much about it beyond that. We can easily have separate horizontal and vertical jitters, and they can be adapted separately (this is well motivated by the fact that buildings are consistently asymmetric).

[the remark this answers is not in the document]
Comment (drh-determined-ai) [reply examples/gan/pix2pix_tf_keras/data.py:77]: I think it's just for simplicity. I imagine the original research was more sophisticated. I think what we have now is good for our example.

File: examples/gan/pix2pix_tf_keras/data.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
+import pathlib
+from typing import Tuple
+
+import tensorflow as tf
+
+
+def download(base, dataset) -> str:
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/gan/pix2pix_tf_keras/data.py:8]: Yep, totally up to you for examples.

File: examples/gan/pix2pix_tf_keras/data.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    )
+    input_image, real_image = _normalize(input_image, real_image)
+
+    return input_image, real_image
+
+
+def load_dataset(path, height, width, set_="train", jitter=0, mirror=False):
+    """Load the images into memory and preprocess them."""
+    ds = tf.data.Dataset.list_files(str(path / f"{set_}/*.jpg"))
+    if set_ != "train":
```
[the remark this answers is not in the document]
Comment (garrett361) [reply examples/gan/pix2pix_tf_keras/data.py:116]: Ok cool, that's what my (?) was about: I wasn't sure how many options were valid for `set_`. Since there are more than two options, disregard this comment.

**Review by azhou-determined — COMMENTED** [review]
can we put the example images in their own directory? it's a bit confusing for someone just reading the code to see all the example png/jpegs

**Review by garrett361 — APPROVED** [review]
